org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/RecyclableObjectPool.java

```diff
@@ -0,0 +1,138 @@
+    /**
+     * Add an object into this pool, if the object is already added, the original id will be used, it will also create a 
+     * reference link from the object to its owner. 
+     * 
+     * @param owner the owner of this object
+     * @param object the object
+     * @return the inner id of this object
+     */
+    public int addObject(O owner, V object) {
+        if (owner == null) {
+            throw new IllegalArgumentException("Owner cannot be null.");
+        }
+        if (object == null) {
+            throw new IllegalArgumentException("Null object cannot be added.");
+        }
+        synchronized (this) {
+            if (!referenceMap.containsKey(object)) {
+                // the object is new
+                Set<O> owners = new HashSet<>(1);
+                owners.add(owner);
+                referenceMap.put(object, owners);
+                int id = objectCollection.create(object);
+                objectIdMap.put(object, id);
+                return id;
+            } else {
+                // the object is already in this pool
+                referenceMap.get(object).add(owner);
+                return objectIdMap.get(object);
+            }
+        }
+    }
+
+    /**
+     * Get the object by object id.
+     * 
+     * @param id the object id.
+     * @return the object, null if the object cannot be found.
+     */
+    public V getObjectById(int id) {
+        synchronized (this) {
+            return objectCollection.get(id);
+        }
+    }
+
+    /**
+     * Remove the object by object id.
+     * 
+     * @param id the object id.
+     * @return true if the object is removed successfully, false if the object cannot be found.
+     */
+    public boolean removeObjectById(int id) {
+        synchronized (this) {
+            V object = this.objectCollection.remove(id);
+            if (object == null)  {
+                return false;
+            }
+            referenceMap.remove(object);
+            objectIdMap.remove(object);
+            return true;
+        }
+    }
+
+    /**
+     * Remove a group of objects with the owner, the objects which only refers this owner will be removed.
+     *  
+     * @param owner the owner.
+     * @return true if any object is removed.
+     */
+    public boolean removeObjectsByOwner(O owner) {
+        if (owner == null) {
+            throw new IllegalArgumentException("owner cannot be null.");
+        }
+        synchronized (this) {
+            List<V> recycling = new ArrayList<>();
+            referenceMap.forEach((key, value) -> {
+                if (value.remove(owner)) {
+                    if (value.isEmpty()) {
+                        recycling.add(key);
+                    }
+                }
+            });
+            for (V recycled : recycling) {
+                this.objectCollection.remove(objectIdMap.remove(recycled));
+                referenceMap.remove(recycled);
+            }
+            return !recycling.isEmpty();
```

Comment:
This makes the function not idempotent. Calling this function twice with the same id will lead to different results. And I don't see the results is checked anywhere.